keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1462,225 +1278,309 @@ prevented. Deleting the entire cluster also leaves resources allocated outside
+    // configuration and requirements.
+    //
+    // +optional
+    DeviceClassName *string
 
-    // Shareable indicates whether the allocated claim is meant to be shareable
-    // by multiple consumers at the same time.
-    Shareable bool
+    // Config defines configuration parameters that apply to the requested resource(s).
+    // They are ignored while allocating the claim.
+    //
+    // +optional
+    // +listType=atomic
+    Config []ConfigurationParameters
+
+    // AdminAccess indicates that this is a claim for administrative access
+    // to the device(s). Claims with AdminAccess are expected to be used for
+    // monitoring or other management services for a device.  They ignore
+    // all ordinary claims to the device with respect to access modes and
+    // any resource allocations. Ability to request this kind of access is
+    // controlled via ResourceQuota in the resource.k8s.io API.
+    //
+    // Default is false.
+    //
+    // +optional
+    AdminAccess *bool
 
-    // DriverRequests describes all resources that are needed for the
-    // allocated claim. A single claim may use resources coming from
-    // different drivers. For each driver, this array has at most one
-    // entry which then may have one or more per-driver requests.
+    // The amount field can be used to ask for all devices which match the
+    // requirements or a certain fixed number. The default is exactly one device.
     //
-    // May be empty, in which case the claim can always be allocated.
-    DriverRequests []DriverRequests
-}
+    // Note that the overall number of allocated devices per claim is limited
+    // to 32, so asking for one device in more than 32 requests and asking for
+    // more than 32 identical devices in a single request are both not allowed.
+    // This gets checked during validation. The number of devices matching
+    // the requirements is not known in advance, so asking for "all devices"
+    // can fail during allocation.
+    //
+    // +optional
+    Amount *Amount
 
-// DriverRequests describes all resources that are needed from one particular driver.
-type DriverRequests struct {
-    // DriverName is the name used by the DRA driver kubelet plugin.
-    DriverName string
+    // Requirements describe additional contraints that all must be met by a device
+    // to satisfy the request.
+    //
+    // +optional
+    // +listType=atomic
+    Requirements []Requirement
+}
 
-    // VendorParameters are arbitrary setup parameters for all requests of the
-    // claim. They are ignored while allocating the claim.
-    VendorParameters runtime.Object
+// Exactly one field must be set.
+type Amount struct {
+    // All, if set, asks for all devices matching the requirements.
+    // Allocation fails if not all of them are available, unless admin
+    // access is requested. Admin access is granted also for
+    // devices which are in use.
+    //
+    // +optional
+    All bool
```

Comment:
I'm not actually sure what semantics `all` should have in the non-admin use-case. I'm specifically thinking of the MIG use-case where the definition of "all" is fuzzy (since it's not actually possible to have access to *all* devices at once).

Reply:
all means "all matching the requirements". So, if you said "all" with a partitionable device you would just never schedule. I am thinking the algorithm is something like:
1. Final all devices on the node matching the requirements. This may include overlapping devices, if the requirements are not refined enough.
2. Evaluate if it is possible to allocate that set; if not, this node is not a fit.

If you wanted all GPUs on a node, you would want to say "non-MIG" in your requirements (or the class would).

Reply:
I think that would work. I couldn't do something like -- give me all GPUs on the node partitioned into X,Y,Z -- but I could actually say -- give me all MIG devices with profile `1g.5gb` and since those are non-overlapping it is actually satisfiable.

Reply:
The device selector would have to limit the request to those devices which actually can be allocated together.

This also would make sense for management access because listing all MIG devices in the results is redundant (some are the same card) and potentially too large (32 limit for results).